core/drivers/stm32_gpio.c

```diff
@@ -77,6 +80,132 @@ static unsigned int gpio_lock;
 static STAILQ_HEAD(, stm32_gpio_bank) bank_list =
 		STAILQ_HEAD_INITIALIZER(bank_list);
 
+static struct stm32_gpio_bank *gpio_chip_to_bank(struct gpio_chip *chip)
+{
+	return container_of(chip, struct stm32_gpio_bank, gpio_chip);
+}
+
+static enum gpio_level stm32_gpio_get_level(struct gpio_chip *chip,
+					    unsigned int gpio_pin)
+{
+	struct stm32_gpio_bank *bank = gpio_chip_to_bank(chip);
+	enum gpio_level level = GPIO_LEVEL_HIGH;
+	unsigned int reg_offset = 0;
+	unsigned int mode = 0;
+
+	assert(gpio_pin < bank->ngpios);
+	clk_enable(bank->clock);
+
+	mode = (io_read32(bank->base + GPIO_MODER_OFFSET) >> (gpio_pin << 1)) &
+	       GPIO_MODE_MASK;
+
+	switch (mode) {
+	case GPIO_MODE_INPUT:
+		reg_offset = GPIO_IDR_OFFSET;
+		break;
+	case GPIO_MODE_OUTPUT:
+		reg_offset = GPIO_ODR_OFFSET;
+		break;
+	default:
```

Comment:
Do we want to panic if we ask a level for these modes?

10: Alternate function mode
11: Analog mode

Reply:
Same as above: a consumer should not get/set a GPIO level if it did not get the GPIO first.